sdk/trace/src/main/scala/org/typelevel/otel4s/sdk/trace/SpanRef.scala

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright 2023 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package sdk
+package trace
+
+import org.typelevel.otel4s.sdk.common.InstrumentationScope
+import org.typelevel.otel4s.sdk.trace.data.SpanData
+import org.typelevel.otel4s.trace.Span
+import org.typelevel.otel4s.trace.SpanContext
+import org.typelevel.otel4s.trace.SpanKind
+
+import scala.concurrent.duration.FiniteDuration
+
+/** An extended Span interface that provides access to internal state.
+  *
+  * Since the span's internal state can be mutated during the lifetime, some
+  * operations are effectful.
+  *
+  * @tparam F
+  *   the higher-kinded type of a polymorphic effect
+  */
+trait SpanRef[F[_]] { self: Span.Backend[F] =>
```

Comment:
Sorry to be late 😅 but I do think we should bikeshed this name. As far as I understand there is nothing `Ref`-like here, i.e. you can't `set` anything. But you may read effectually, which does make it similar to `RefSource`.

Sorry, I don't have better naming ideas 😕

Reply:
Java SDK uses ReadWriteSpan and ReadableSpan. 
ReadWriteSpan is not ideal, but more meaningful